build-data-capturing-gradle-samples/capture-git-diffs/gradle-git-diffs.gradle

```diff
@@ -21,80 +21,102 @@ if (!buildScanApi) {
-                }
+    void captureGitDiffInGist(api) {
+        def hasCredentials = gistTokenProvider.isPresent()
+        if (!hasCredentials) {
+            logger.warn("User has not set 'gistToken'. Cannot publish gist.")
+            return
+        }
+
+        def diff = execAndGetStdout('git', 'diff')
+        if (diff) {
+            try {
+                def baseUrl = new URL('https://api.github.com/gists')
+                def basicAuth = "Basic ${gistTokenProvider.get().bytes.encodeBase64()}"
 
-                // response
-                def url = new JsonSlurper().parse(content.text.bytes).html_url
-                buildScan.link('Git diff', url)
+                HttpURLConnection connection = (HttpURLConnection) baseUrl.openConnection()
+                connection.with {
+                    // request
+                    setRequestProperty('Authorization', basicAuth)
+                    setRequestProperty('Accept', 'application/vnd.github.v3+json')
+                    requestMethod = 'POST'
+                    doOutput = true
+                    outputStream.withWriter { writer ->
+                        jsonRequest(writer, diff, projectName)
+                    }
+
+                    // response
+                    def url = new JsonSlurper().parse(content.text.bytes).html_url
+                    api.link('Git diff', url)
+                }
+                logger.info('Successfully published gist.')
+            } catch (ex) {
+                logger.warn('Unable to publish gist', ex)
             }
-            gradle.rootProject.logger.info('Successfully published gist.')
-        } catch (ex) {
-            gradle.rootProject.logger.warn('Unable to publish gist', ex)
         }
     }
-}
 
-// this method must be static, otherwise Gradle will interpret `files` as Project.files() and this won't work
-static void jsonRequest(Writer writer, String diff, Project project) {
-    def builder = new JsonBuilder()
-    builder {
-        description "Git diff for $project.name"
-        'public' false
-        files {
-            "${project.name}.diff" {
-                content diff
+    void jsonRequest(Writer writer, String diff, String projectName) {
+        def builder = new JsonBuilder()
+        builder {
+            description "Git diff for $projectName"
+            'public' false
+            files {
+                "${projectName}.diff" {
+                    content diff
+                }
             }
         }
+        builder.writeTo(writer)
     }
-    builder.writeTo(writer)
-}
 
-static String execAndGetStdout(String... args) {
-    Process process = args.toList().execute()
-    try {
-        def standardText = process.inputStream.withStream { s -> s.getText(Charset.defaultCharset().name()) }
-        def ignore = process.errorStream.withStream { s -> s.getText(Charset.defaultCharset().name()) }
+    String execAndGetStdout(String... args) {
```

Comment:
can these functions be kept static and outside the class or is this required for configuration caching?

Reply:
You need the method `captureGitDiffInGist` to be encapsulated.
Then it would be awkward to me to scatter the logic in/out of the `Capture` class, but if following this approach, how would you refer to the static method, say `execAndGetStdout` in the root scope?
The `project` is not accessible from within the `Capture` class.

Reply:
ah i see. let's keep it here then. this will minimize changes to the ordering and structure of the code.